### ISSUE TITLE
Open source alternative?

### ISSUE THREAD

**Issue by z3t0** [screeps/screeps#53]
I was wondering if there was an opensource game like screeps? I realize the server is open source, but I mean the entire game

**Comment by daboross** [comment]
I believe the developers do not intend to open source the official client, but there are multiple users in the community who are working on rebuilding an open source client from scratch.

The protocol with which the Screeps server talks to is available in this repository and via browser debug tools, the largest challenge really is just the client graphics.

**Comment by tedivm** [comment]
Personally I'm hoping the developers will put out another crowd funding campaign to open source the new front end client they are working on, but they haven't said anything that would make me believe this is being planned.

**Comment by z3t0** [comment]
@daboross @tedivm I would really want an open source game very much that follows the same principles. I will begin working on this and would be thrilled if anyone else wanted to join?

The main reason I say so, is that a game based on extensibility and programming really should be open source because devs are not going to want to wait on the producers to implement changes, what if I want to add support for a built in emscripten compiler? What if I want to be able to connect to a local server from the browser? Maybe I want to change the way CPU ticks are counted etc.

**Comment by daboross** [comment]
@z3t0 I'm currently working on a desktop native client, the typed protocol interpretation I've done so far is at github.com/daboross/rust-screeps-api/.

I'm mostly working on a desktop app so that screeps can be run on older computers efficiently, since the web app (and steam client which is really just the web app in electron) slows considerably with more than one tab/window open. Having at open source client is not really my *main* reason for this, but it will be open source, and will be usable if I ever get all functionality finished.

What sort of technology / target would you be aiming for? I think having an open source browser-based client would be pretty awesome, but I know next-to-no JavaScript to implement anything with.

**Comment by ags131** [comment]
@z3t0 A vast majority of the game is already open source, and some of us have made quite a few mods and forks. For example, I have a fork that modifies the server to run on mongo and redis, and also have a WIP opensource client using electron that is currently capable of displaying a room and visuals. I'm using WebGL for it to avoid the memory heavy DOM the official client uses. 
This client once complete will be able to run both standalone and in the browser and support both official and private servers

**Comment by tedivm** [comment]
> The main reason I say so, is that a game based on extensibility and programming really should be open source because devs are not going to want to wait on the producers to implement changes, what if I want to add support for a built in emscripten compiler? What if I want to be able to connect to a local server from the browser? Maybe I want to change the way CPU ticks are counted etc.

There's nothing stopping you from doing all of that now. The server is open source, which means you can add your own custom language stuff in (one person even made a special server mode so that "remote servers" can send "intent" to the server directly, bypassing language completely). You can easily change the way CPU ticks are counted on your own server already. The protocol to talk with servers is open, so you can even make your own browser client.

**Comment by daboross** [comment]
I guess I responded to the first part of the message, but for the second part, I'd aree with tedivm. Everything you mentioned besides connecting via the browser to a local server is currently possible, even pretty easy to do with the server modding and module framework.

Making a custom code backend requires no client modifications, all you'd need to do is fork the 'runner' server module, modify it to run the language you want, and make some way to handle intents in that language. The server structure is completely modular, so you could even rewrite the runner in your language of choice, provided you implemented intent communication with the other modules correctly.

**Comment by ags131** [comment]
@daboross Even connecting to a local server via browser is possible as long as the server has screepsmod-auth (OR another auth mod) installed. My client currently does this :)

**Comment by daboross** [comment]
True, though not using the official client without modification

**Comment by z3t0** [comment]
> There's nothing stopping you from doing all of that now. The server is open source, which means you can add your own custom language stuff in (one person even made a special server mode so that "remote servers" can send "intent" to the server directly, bypassing language completely). You can easily change the way CPU ticks are counted on your own server already. The protocol to talk with servers is open, so you can even make your own browser client.


Thank you for explaining that? I realized the server was open source but not that the protocol to communicate was... just to be clear then, the entire game is essentially open source? I mean besides the official client but that's not really very important anyways.

**Comment by daboross** [comment]
It really does depend on your definition of "entire game".

The protocol the client talks to the server with is definitely "open", and it is not that complicated - mostly just nicely laid out REST calls with a websocket connection for ongoing updates.

But there are things that are only in the "official client". For example, none of the SVG images for game objects are open source, and are only contained in the client. The code to render different badge types is also similarly closed source (server just returns an integer badge ID and colors for each badge, not a full image).

These things can be replicated, but they definitely aren't open source to begin with.

if you don't refer to the client but rather the game logic though, yes, all of it is in this (and related) open source server repositories.

**Comment by tedivm** [comment]
It isn't all very well documented but since the server- which defines and implements the protocol- is open source than by definition so is the protocol. There are a *ton* of third party extensions that take advantage of this. In the [screepers org](https://github.com/screepers) you can see quite a few of them, as well as the unofficial javascript and python api clients (I maintain the python one, and it's pretty up to date).

The one thing to be mindful of is that the developers can change the API at any time without notice. In practice I haven't seen this happen though- they've certainly added to it but there have been very few backwards incompatible changes.

@daboross - this may be tangential but a couple of people have independently written and open sourced badge renderers. This is a good example of how things that aren't open yet tend to get built eventually by the community, although an open source front end client is a rather large undertaking so it's taking some time.

**Comment by daboross** [comment]
For the badges, I've seen screepspl.us - it copied the SVG path data for each badge type directly from the client JS though doesn't it? I mean it's still pretty nice to have that as an open service, but I don't think it is exactly "rebuilt" logic.

For something as specific as the way to render badges whose shapes screeps made up, I think it would be quite hard to re-implement without copying at least some of the SVG data from the client.

In any case, things definitely can be built by the community! I was just bringing up badges mostly as something that has only been officially published within the client, and is still mostly part of the game.

**Comment by tedivm** [comment]
I've seen two different libraries that handle it, but you are right that I can't verify the level of clean rooming they did during development.

On the plus side it means I can make [services](http://www.leagueofautomatednations.com/badges/) that do [this](http://www.leagueofautomatednations.com/badges/alliances/daboross.png)

**Comment by z3t0** [comment]
Is there any way to play on as private server without purchasing the game.

**Comment by daboross** [comment]
Not with an actual game client, no. Once any of the open source clients mentioned are complete, it will be possible, but currently you either need to buy the $15 client, or only interact with the world though HTTP requests xD.

In all seriousness, while I am very glad they open sourced the server, a lot of the game for me is the community that only (right now) exists on the official world. I'm glad that this server and the upcoming OS clients will let more people play the game, but I still think a lot of the value is in the awesome communities than have built up in the different regions of the official server.

**Comment by z3t0** [comment]
I absolutely agree! I personally purchased the game as well. The reason I ask is that I run a coding club at my high school and wanted to introduce Screeps as a way to teach javascript. The main problem is that I do not think purchasing the game for each student is a reasonable request to admin... 

> Once any of the open source clients mentioned are complete, it will be possible, but currently you either need to buy the $15 client, or only interact with the world though HTTP requests xD.

Can you link me to some of the most complete clients? Maybe I'll just direct my efforts there.

**Comment by tedivm** [comment]
You should join the slack network and head into #servers to chat with some of the client developers.

**Comment by ags131** [comment]
I've got a partial one at https://github.com/ags131/screeps-client, room rendering is nearly complete, map view and UI are all missing. As @tedivm mentioned, there are quite a few of us in slack working on clients, best bet is to hop in there and ask. :)

**Comment by daboross** [comment]
Ags's client is much further along than mine, but I've got some work on a native client at https://github.com/daboross/rust-screeps-api and https://github.com/daboross/conrod-testing. 

It has rust typed bindings to many of the API endpoints, a basic framework for the UI and a framework for threaded network calls and caching network results. I have no work done on rendering yet, but I'm working towards that! Right now all it does is let you log in, and it displays your username and GCL.

**Comment by z3t0** [comment]
Thank you so much for all the information! I'll go do that then.

**Comment by z3t0** [comment]
I took a look at @ags131 client but it seems that it's based on electron? I was hoping to work on / create a client that would have a main focus on performance and efficiency. And from my experience electron is not the way to go.

@daboross Would you be open to discussing me contributing to your client?

**Comment by daboross** [comment]
I do think ags131's client has a focus on more performance as well, at least more than the official client, as it's using WebGL (rather than dom rendering... I think?). In any case he can probably speak to it better, I don't know too much about web technologies.

@z3t0 sure though - my client's definitely in prototyping stages, we can discuss that in #servers now that you're there!

**Comment by cathalgarvey** [comment]
Ah, this is a shame. I was mislead into thinking that Screeps was an open source game, and wasted a bunch of time learning to play. I was confused at first when I saw that Steam was required even to run a "local server", now I see that you can't actually play it without a proprietary client and a Steam DRM account.

I hope this changes someday, it looks like a cool game.

**Comment by artch** [comment]
@cathalgarvey Steam is not required to run a local server, it is fully open sourced and can be launched standalone, see the instructions in README in this repo.

The official GUI client is proprietary, this is correct.

**Comment by cathalgarvey** [comment]
Thanks for the clarification; without a worthy GUI I won't be investigating the server any further. Good luck, I hope things change on this front someday.

**Comment by artch** [comment]
We don't have any plans to create a free client due to our business model, but there is a number of ongoing community projects regarding this.

**Comment by cathalgarvey** [comment]
I didn't suggest making it free, only (at least) open source. I didn't expect ye to change this stance, though; few companies that have adopted it ever do. I've noticed the efforts from other developers, good luck to them.
